### PR TITLE
Add service summary metrics aggregator

### PR DESCRIPTION
**Pull request by carsonip**
## Motivation/summary

Add a service summary metrics aggregator which aggregates `agent.name`, `service.environment`, `service.name`, `labels` and `numeric_labels` over configured intervals from all events except spans. The aggregation writes to a new data stream `metrics-apm.service_summary.${interval}` with intervals `1m`, `10m` and `60m`.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [x] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [x] Documentation has been updated

## How to test these changes

- Ensure service_summary data stream has the aggregated document on receiving any events

## Related issues

Part of #9654

### CODE REVIEW

File: x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+	return a.config.BatchProcessor.ProcessBatch(ctx, &batch)
+}
+
+// ProcessBatch aggregates all service summary metrics.
+func (a *Aggregator) ProcessBatch(ctx context.Context, b *model.Batch) error {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	for _, event := range *b {
+		// Not filtering on Processor.
+		a.processEvent(&event, b)
```
Comment (carsonip) [inline x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go:149]: Are there any events that we need to exclude? There is absolutely 0 filtering now

File: x-pack/apm-server/main.go [hunk 2]
```diff
@@ -150,6 +151,19 @@ func newProcessors(args beater.ServerParams) ([]namedProcessor, error) {
...
 		processors = append(processors, namedProcessor{name: serviceName, processor: serviceAggregator})
+
+		const serviceSummaryName = "service summary aggregation"
+		args.Logger.Infof("creating %s with config: %+v", serviceSummaryName, args.Config.Aggregation.Service)
+		serviceSummaryAggregator, err := servicesummarymetrics.NewAggregator(servicesummarymetrics.AggregatorConfig{
+			BatchProcessor:  args.BatchProcessor,
+			Interval:        metricsInterval,
+			RollUpIntervals: rollUpMetricsIntervals,
+			MaxGroups:       args.Config.Aggregation.Service.MaxGroups,
+		})
```
Comment (carsonip) [inline x-pack/apm-server/main.go:162]: Reusing service metrics config `enabled` and `max_groups` here. Is this fine or should we introduce a new set of config for service summary?

File: x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+		if !a.active[period].storeOrUpdate(key, a.config.Logger) {
+			intervalStr := interval.FormatDuration(period)
+			*b = append(*b, makeMetricset(key, intervalStr))
+		}
+	}
+}
+
+type metricsBuffer struct {
+	mu sync.RWMutex
+	m  map[uint64]aggregationKey
```
Comment (carsonip) [inline x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go:166]: pending discussion on hash collision and whether to use a chained hash table like service metrics

File: x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go (hunk 1 above)
Comment (simitt) [reply x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go:149]: strictly thinking we could ignore spans, they are only adding noise and no value, but they do not increase cardinality, so also not a big issue to not have a filter at all

File: x-pack/apm-server/main.go [hunk 4]
```diff
@@ -104,7 +105,7 @@ type processor interface {
 // newProcessors returns a list of processors which will process
 // events in sequential order, prior to the events being published.
 func newProcessors(args beater.ServerParams) ([]namedProcessor, error) {
-	processors := make([]namedProcessor, 0, 3)
+	processors := make([]namedProcessor, 0, 5)
```
Comment (simitt) [inline x-pack/apm-server/main.go:108]: nice catch

File: x-pack/apm-server/main.go (hunk 2 above)
Comment (simitt) [reply x-pack/apm-server/main.go:162]: clarified on the issue - fine to reuse for now

**Review by simitt — COMMENTED** [review]
PR generally looks good to me. There is no overflow handling if the max groups limit is exceeded, @dgieselaar @felixbarny would you expect this to be aligned with the newly introduced overflowbucket `_other` for transactions and service transaction metrics? (would that even be useful for the UI)?

**Comment by simitt** [comment]
@carsonip can you add the logic for the overflow bucket to this PR?

File: x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go (hunk 3 above)
Comment (axw) [reply x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go:169]: We definitely need to handle collisions. How that's implemented is less important, but we should probably be consistent and use separate chaining.

**Comment by simitt** [comment]
@carsonip having an indicator without an actual aggregation is fine for now, as the next step will be to add the counters per event type.

**Comment by carsonip** [comment]
> @carsonip having an indicator without an actual aggregation is fine for now, as the next step will be to add the counters per event type. I'd also be fine to do this in a dedicated PR though, as you are right that it makes much more sense once the counters are added.

In that case, let's do it in a dedicated PR. Currently service summary will add the event to the batch without aggregating on overflow, similar to what's in service destination metrics.

File: x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go (hunk 1 above)
Comment (carsonip) [reply x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go:154]: Done, thanks.

**Review by simitt — APPROVED** [review]
Changes look good. 

Please note on the related issue which parts are implemented with this PR and which parts are left open.

**Pull request merged**